ospfd/ospf_sr.c

```diff
@@ -1115,38 +1149,85 @@ void ospf_sr_ri_lsa_update(struct ospf_lsa *lsa)
-			 __func__);
-		hash_release(OspfSR.neighbors, &(srn->adv_router));
-		XFREE(MTYPE_OSPF_SR_PARAMS, srn);
+	/* Check if Segment Routing Capabilities has been found */
+	if (ri_srgb == NULL) {
+		/* Skip Router Information without SR capabilities
+		 * advertise by a non SR Node */
+		if (srn == NULL) {
+			return;
+		} else {
+			/* Remove SR Node that advertise Router Information
+			 * without SR capabilities. This could correspond to a
+			 * Node stopping Segment Routing */
+			hash_release(OspfSR.neighbors, &(srn->adv_router));
+			sr_node_del(srn);
+			return;
+		}
+	}
+
+	/* Check that RI LSA belongs to the correct SR Node */
+	if ((srn != NULL) && (srn->instance != 0)
+	    && (srn->instance != ntohl(lsah->id.s_addr))) {
+		flog_err(EC_OSPF_SR_INVALID_LSA_ID,
+			 "SR (%s): Abort! Wrong "
+			 "LSA ID 4.0.0.%u for SR node %pI4/%u",
+			 __func__, GET_OPAQUE_ID(ntohl(lsah->id.s_addr)),
+			 &lsah->adv_router, srn->instance);
 		return;
 	}
 
+	/* OK. All things look good. Get SRGB */
+	srgb.range_size = GET_RANGE_SIZE(ntohl(ri_srgb->size));
+	srgb.lower_bound = GET_LABEL(ntohl(ri_srgb->lower.value));
+
 	/* Check if it is a new SR Node or not */
-	if (srn->instance == 0) {
+	if (srn == NULL) {
+		/* Get a new SR Node in hash table from Router ID */
+		srn = (struct sr_node *)hash_get(OspfSR.neighbors,
+						 (void *)&(lsah->adv_router),
+						 (void *)sr_node_new);
+		/* Sanity check */
+		if (srn == NULL) {
+			flog_err(
+				EC_OSPF_SR_NODE_CREATE,
+				"SR (%s): Abort! can't create SR node in hash table",
+				__func__);
+			return;
+		}
 		/* update LSA ID */
 		srn->instance = ntohl(lsah->id.s_addr);
 		/* Copy SRGB */
 		srn->srgb.range_size = srgb.range_size;
 		srn->srgb.lower_bound = srgb.lower_bound;
+		/* Set Algorithm */
+		if (algo != NULL) {
+			int i;
+			for (i = 0; i < ntohs(algo->header.length); i++)
+				srn->algo[i] = algo->value[0];
+			for (; i < ALGORITHM_COUNT; i++)
+				srn->algo[i] = SR_ALGORITHM_UNSET;
+		} else {
+			srn->algo[0] = SR_ALGORITHM_SPF;
+		}
+		/* set MSD */
+		srn->msd = msd;
+		return;
 	}
 
 	/* Check if SRGB has changed */
-	if ((srn->srgb.range_size != srgb.range_size)
-	    || (srn->srgb.lower_bound != srgb.lower_bound)) {
-		srn->srgb.range_size = srgb.range_size;
-		srn->srgb.lower_bound = srgb.lower_bound;
-		/* Update NHLFE if it is a neighbor SR node */
-		if (srn->neighbor == OspfSR.self)
-			hash_iterate(OspfSR.neighbors,
-				     (void (*)(struct hash_bucket *,
-					       void *))update_out_nhlfe,
-				     (void *)srn);
-	}
+	if ((srn->srgb.range_size == srgb.range_size)
+	    && (srn->srgb.lower_bound == srgb.lower_bound))
+		return;
+
+	/* Update SRGB ... */
+	srn->srgb.range_size = srgb.range_size;
+	srn->srgb.lower_bound = srgb.lower_bound;
+	/* ... and NHLFE if it is a neighbor SR node */
+	if (srn->neighbor == OspfSR.self)
+		hash_iterate(OspfSR.neighbors,
+			     (void (*)(struct hash_bucket *,
+				       void *))update_out_nhlfe,
+			     (void *)srn);
```

Comment:
nit: unnecessary explicit cast to `void *`

Reply:
done